sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4

```diff
@@ -354,9 +355,14 @@ resource
     : identifier STRING
     ;
 
+insertStatement
+    : (ctes)? insertInto queryTerm queryOrganization                                       #singleInsertQuery
+    | (ctes)? fromClause multiInsertQueryBody+                                             #multiInsertQuery
+    ;
+
 queryNoWith
-    : insertInto? queryTerm queryOrganization                                              #singleInsertQuery
-    | fromClause multiInsertQueryBody+                                                     #multiInsertQuery
+    : queryTerm queryOrganization                                               #noWithQuery
+    | fromClause querySpecification queryOrganization                           #queryWithFrom
```

Comment:
where does this come from?

Reply:
@cloud-fan This is required to parse statement of the form [code](https://github.com/apache/spark/blob/e402de5fd030cdc4150fda0755c7c636cad9619e/sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala#L1007=L1009)

Reply:
Why isn't it covered by `insertStatement`? We have `(ctes)? fromClause multiInsertQueryBody+` there

Reply:
@cloud-fan Actually this is sort of tricky..  I initially didn't have this rule thinking insertStatement should take care of this. But it didn't .
```
FROM (FROM TEST SELECT TRANSFORM(KEY, VALUE) USING 'CAT' AS (`THING1` INT, THING2 STRING)) T SELECT THING1 + 1
```
In this case, we have nested from clause. and the inner from is handled by rule `aliasedQuery`. If we see the definition, it calls `queryNoWith` and we resolve the nested FROM with along with querySpecification(the transform query) by using queryNoWith:queryWithFrom rule.